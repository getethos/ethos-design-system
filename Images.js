import React from 'react'
import PropTypes from 'prop-types'
import * as Cloudinary from 'cloudinary-react'

// https://cloudinary.com/documentation/image_transformation_reference

export const CLOUDINARY_CLOUD_NAME = 'ethos-media'

const publicIdBase = `https://res.cloudinary.com/${CLOUDINARY_CLOUD_NAME}/image/upload/`

export const CloudinaryImage = ({ publicId, ...rest }) => {
  const publicIdFilename = publicId.replace(publicIdBase, '')
  return (
    <Cloudinary.Image
      {...rest}
      className="Image"
      cloudName={CLOUDINARY_CLOUD_NAME}
      dpr={2} // should optimize
      publicId={publicIdFilename}
      crop="fill" // aka CSS cover
      secure="true"
    />
  )
}

CloudinaryImage.propTypes = {
  publicId: PropTypes.string.isRequired,
}
