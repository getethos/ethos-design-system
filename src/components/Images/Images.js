import React from 'react'
import PropTypes from 'prop-types'
import * as Cloudinary from 'cloudinary-react'

// https://cloudinary.com/documentation/image_transformation_reference

export const CLOUDINARY_CLOUD_NAME = 'getethos'

const publicIdBase = `https://res.cloudinary.com/${CLOUDINARY_CLOUD_NAME}/image/upload/`

export const CloudinaryImage = ({ publicId, className, ...rest }) => {
  const publicIdFilename = publicId.replace(publicIdBase, '')
  return (
    <Cloudinary.Image
      {...rest}
      className={['Image', className].join(' ')}
      cloudName={CLOUDINARY_CLOUD_NAME}
      dpr="auto"
      responsive
      publicId={publicIdFilename}
      crop="fill" // aka CSS cover
      secure="true"
    />
  )
}

CloudinaryImage.propTypes = {
  // We're trying to avoid/reduce className restyling, but let's try it here:
  className: PropTypes.string,
  publicId: PropTypes.string.isRequired,
}
