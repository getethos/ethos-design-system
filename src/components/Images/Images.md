```jsx
<div style={{ width: 650 }}>
  <ul>
    <li>These images should lazy load.</li>
    <li>
      When viewed in a retina screen (like a macbook) this will load a high
      resolution image.
    </li>
    <li>
      You can test this image's responsiveness with chrome inspector using an
      ipad, an iphone 6/7/8, and an iphone 5/SE; they should load different
      images based on device resolution (although the specifics are determined
      by ) and hard refreshing with different devices.
    </li>
  </ul>
  <CloudinaryImage
    width={650}
    alt="father and kids playing"
    publicId="v1565712179/01_NEW%20Lifestyle%20%28Rebrand%29/life-insurance-father-and-kids-playing.jpg"
  />
</div>

<br/>
<br/>

<ul>
  <li>Due to not being constrained by the `width` property, this image should be bigger. It should also load a very large, retina-resolution image in comparison to the above image.
  </li>
  <li>
    Without getting passed in a "width" prop, this image will be higher resolution on all retina displays, despite sometimes displaying smaller. I think this is OK, although it seems off?
  </li>
</ul>
<CloudinaryImage
  alt="father and kids playing"
  publicId="v1565712179/01_NEW%20Lifestyle%20%28Rebrand%29/life-insurance-father-and-kids-playing.jpg"
/>
```
