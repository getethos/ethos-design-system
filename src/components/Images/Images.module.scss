/* @getethos/design-system/Images.scss
   ========================================================================== */

/* WIP. */

/* Generally used as a wrapper around Image components. */
figure {
  margin: 0;
}

.Image {
  display: block;
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.Svg {
  display: block;
  width: 100%;
  height: auto;
}

.blurUp {
  filter: blur(5px);
  transition: filter 500ms;

  &:global(.lazyloaded) {
    filter: blur(0);
  }
}
