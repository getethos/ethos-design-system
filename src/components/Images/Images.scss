/* @getethos/design-system/Images.scss
   ========================================================================== */

// WIP.

/* Generally used as a wrapper around Image components. */
figure {
  margin: 0;
}

.Image {
  // border-radius: 2px; removed until reason for this being added is confirmed
  display: block;

  // These are consequential, may need to pull out.
  width: 100%;
  height: auto;
}
