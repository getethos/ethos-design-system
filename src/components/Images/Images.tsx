import cloudinary, { Cloudinary } from 'cloudinary-core'
import * as PropTypes from 'prop-types'
import * as React from 'react'
import { v4 as uuidv4 } from 'uuid'
import useInvalid from '../../hooks/useInvalid.js'
import useRequired from '../../hooks/useRequired.js'
import { Media } from '../Media/Media'
import styles from './Images.module.scss'
// https://cloudinary.com/documentation/image_transformation_reference
export const CLOUDINARY_CLOUD_NAME = 'getethos'
// @ts-ignore
const cld: Cloudinary = new cloudinary.Cloudinary({
  cloud_name: CLOUDINARY_CLOUD_NAME,
})
const mediaBreakpoints = [
  Media.BREAKPOINTS.DESKTOP_RANGE_START,
  Media.BREAKPOINTS.LAPTOP_RANGE_START,
  Media.BREAKPOINTS.TABLET_RANGE_START,
  Media.BREAKPOINTS.PHONE_RANGE_END,
]
type CloudinaryImageProps = {
  height?: any[]
  width?: any[]
  className?: string
  alt?: string
  publicId: string
  crop?: any
}
const CROP_METHODS = {
  FILL: 'fill',
  FIT: 'fit',
  CROP: 'crop',
}
const PUBLIC_PROPS = {
  height: PropTypes.array,
  width: PropTypes.array,
  className: PropTypes.string,
  alt: PropTypes.string,
  publicId: PropTypes.string.isRequired,
  crop: PropTypes.oneOf(Object.values(CROP_METHODS)),
}
export const CloudinaryImage: React.FC<CloudinaryImageProps> & {
  CROP_METHODS: typeof CROP_METHODS
  PUBLIC_PROPS: typeof PUBLIC_PROPS
} = ({
  publicId,
  className,
  alt = '',
  width,
  height,
  crop = CROP_METHODS.FILL,
  ...rest
}) => {
  // Verify that all required props were supplied
  const [includesRequired] = useRequired(['publicId', 'alt'])
  const allRelevantProps = Object.assign({}, rest, {
    publicId,
    alt,
    width,
    height,
  })
  includesRequired(allRelevantProps)
  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(Object.keys(PUBLIC_PROPS))
  includesInvalid(rest)
  const baseImageSettings = {
    quality: 'auto:eco',
    crop,
    secure: true,
    fetchFormat: 'auto',
    flags: ['progressive:semi'],
  }
  let imageClasses = ['lazyload', className]
  let reverseWidth, reverseHeight
  if (width) {
    reverseWidth = width.slice().reverse()
  }
  if (height) {
    reverseHeight = height.slice().reverse()
  }
  const buildImageTag = (srcSet: string[]) => {
    const srcSetString = srcSet
      .map((url, index) => {
        if (index === 0) return `${url} ${mediaBreakpoints[index] + 1}w`
        return `${url} ${mediaBreakpoints[index - 1]}w`
      })
      .reverse()
      .join(', ')
    // srcSet is the LQIP image, src is the fallback image for older browsers
    // that do not support the 'srcSet' attribute (IE zero support)
    return (
      <img
        key={`${uuidv4()}`}
        className={[styles.Image, styles['blurUp'], ...imageClasses].join(' ')}
        src={cld.url(filePath(publicId), {
          transformation: 'unsupported',
          ...baseImageSettings,
        })}
        srcSet={srcSetString}
        alt={alt}
      />
    )
  }
  const buildTags = () => {
    const dprSettings = ['1.0', '2.0', '3.0']
    let tags: JSX.Element[] = []
    let imageSrcSet: string[] = []
    // We are expecting width/height attribute arrays in the order of
    // Phone/Tablet/Laptop/Desktop but we have to setup media queries
    // in the opposite order, so we reverse the arrays here.
    for (
      let breakpoint = 0;
      breakpoint < mediaBreakpoints.length;
      breakpoint++
    ) {
      const imageSettings = {
        ...baseImageSettings,
        ...(reverseWidth &&
          !!reverseWidth[breakpoint] && { width: reverseWidth[breakpoint] }),
        ...(reverseHeight &&
          !!reverseHeight[breakpoint] && { height: reverseHeight[breakpoint] }),
      }
      const srcsetData = dprSettings.map((dpr, indx) => {
        const sourceSettings = {
          ...imageSettings,
          dpr,
        }
        return cld.url(filePath(publicId), sourceSettings) + ` ${indx + 1}x`
      })
      const minMax = breakpoint < mediaBreakpoints.length - 1 ? `min` : `max`
      tags.push(
        <source
          key={`${uuidv4()}`}
          media={`(${minMax}-width: ${mediaBreakpoints[breakpoint]}px)`}
          data-srcset={srcsetData.join(', ')}
        />
      )
      // ethan - cloudinary typings are buggy so ignoring
      const urlWithChainedTransformation = cld
        .imageTag(filePath(publicId), imageSettings)
        .transformation()
        .chain()
        .transformation('lqip')
        // @ts-ignore
        .getParent()
        .getAttr('src') as string
      imageSrcSet.push(urlWithChainedTransformation)
    }
    tags.push(buildImageTag(imageSrcSet))
    return tags
  }
  const isSvg = (publicId) => {
    return publicId.split('.').pop() === 'svg'
  }
  const renderSvg = () => {
    const baseSvgSettings = {
      secure: true,
    }
    const svgUrl = cld.url(filePath(publicId), baseSvgSettings)
    return (
      <img
        data-src={svgUrl}
        className={[styles.Svg, ...imageClasses].join(' ')}
        alt={alt}
      />
    )
  }
  // Serve a simpler version if resource is SVG
  if (isSvg(publicId)) return renderSvg()
  return <picture>{buildTags()}</picture>
}
export const filePath = (publicId) => {
  if (!publicId) return ''
  const publicIdBase = `https://res.cloudinary.com/${CLOUDINARY_CLOUD_NAME}/image/upload/`
  return publicId.replace(publicIdBase, '')
}
CloudinaryImage.CROP_METHODS = CROP_METHODS
CloudinaryImage.PUBLIC_PROPS = PUBLIC_PROPS
