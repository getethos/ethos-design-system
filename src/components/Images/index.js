export {
  CloudinaryImage,
  CLOUDINARY_CLOUD_NAME,
  PreloadImageTags,
} from './Images'
