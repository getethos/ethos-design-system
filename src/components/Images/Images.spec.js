import React from 'react'
import { CloudinaryImage, preloadImageData } from './Images.js'
import * as Images from './Images'
import renderer from 'react-test-renderer'

describe('CloudinaryImage', () => {
  describe('API', () => {
    test('exports properly', () => {
      expect(CloudinaryImage).toBeDefined()
      expect(CloudinaryImage.CROP_METHODS).toBeDefined()
      expect(Images.CLOUDINARY_CLOUD_NAME).toBeDefined()
    })
  })

  describe('default rendering', () => {
    test('CloudinaryImage', () => {
      const tree = renderer
        .create(
          <CloudinaryImage
            publicId="something.com/otherthing.png"
            className="testImage"
            alt="alt text"
            width={[100, 200, 300, 400]}
            height={[100, 200, 300, 400]}
          />
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('non-lazy rendering, no data-srcset', () => {
    test('CloudinaryImage', () => {
      const tree = renderer
        .create(
          <CloudinaryImage
            publicId="something.com/otherthing.png"
            className="testImage"
            alt="alt text"
            width={[100, 200, 300, 400]}
            height={[100, 200, 300, 400]}
            lazyLoad={false}
          />
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    test('filePath', () => {
      expect(Images.CLOUDINARY_CLOUD_NAME).toBe('getethos')
      const path = Images.filePath(
        'https://res.cloudinary.com/getethos/image/upload/test/image.png'
      )
      expect(path).toBe('test/image.png')
    })
  })
})

describe('preloadImageData', () => {
  describe('API', () => {
    test('exports properly', () => {
      expect(preloadImageData).toBeDefined()
    })
  })

  describe('default rendering', () => {
    test('preloadImageData', () => {
      expect(
        preloadImageData({
          publicId: 'something.com/otherthing.png',
          crop: CloudinaryImage.CROP_METHODS.FILL,
          width: [100, 200, 300, 400],
          height: [100, 200, 300, 400],
        })
      ).toEqual([
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/v1/something.com/otherthing.png',
          as: 'image',
          media: '(max-width: 599px) and (-webkit-max-device-pixel-ratio: 1.0)',
          key: '0-1.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(max-width: 599px) and (-webkit-min-device-pixel-ratio: 1.1) and (-webkit-max-device-pixel-ratio: 2.0)',
          key: '0-2.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(max-width: 599px) and (-webkit-min-device-pixel-ratio: 2.1) and (-webkit-max-device-pixel-ratio: 3.0)',
          key: '0-3.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 600px) and (max-width: 899px) and (-webkit-max-device-pixel-ratio: 1.0)',
          key: '1-1.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 600px) and (max-width: 899px) and (-webkit-min-device-pixel-ratio: 1.1) and (-webkit-max-device-pixel-ratio: 2.0)',
          key: '1-2.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 600px) and (max-width: 899px) and (-webkit-min-device-pixel-ratio: 2.1) and (-webkit-max-device-pixel-ratio: 3.0)',
          key: '1-3.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 900px) and (max-width: 1199px) and (-webkit-max-device-pixel-ratio: 1.0)',
          key: '2-1.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 900px) and (max-width: 1199px) and (-webkit-min-device-pixel-ratio: 1.1) and (-webkit-max-device-pixel-ratio: 2.0)',
          key: '2-2.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 900px) and (max-width: 1199px) and (-webkit-min-device-pixel-ratio: 2.1) and (-webkit-max-device-pixel-ratio: 3.0)',
          key: '2-3.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 1200px) and (-webkit-max-device-pixel-ratio: 1.0)',
          key: '3-1.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 1200px) and (-webkit-min-device-pixel-ratio: 1.1) and (-webkit-max-device-pixel-ratio: 2.0)',
          key: '3-2.0',
        },
        {
          rel: 'preload',
          href:
            'https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/v1/something.com/otherthing.png',
          as: 'image',
          media:
            '(min-width: 1200px) and (-webkit-min-device-pixel-ratio: 2.1) and (-webkit-max-device-pixel-ratio: 3.0)',
          key: '3-3.0',
        },
      ])
    })
  })
})
