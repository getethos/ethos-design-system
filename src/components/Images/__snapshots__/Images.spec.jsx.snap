// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudinaryImage default rendering CloudinaryImage 1`] = `
<picture>
  <source
    data-srcset="https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/v1/something.com/otherthing.png 1x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/v1/something.com/otherthing.png 2x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/v1/something.com/otherthing.png 3x"
    media="(min-width: 1200px)"
  />
  <source
    data-srcset="https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/v1/something.com/otherthing.png 1x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/v1/something.com/otherthing.png 2x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/v1/something.com/otherthing.png 3x"
    media="(min-width: 900px)"
  />
  <source
    data-srcset="https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/v1/something.com/otherthing.png 1x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/v1/something.com/otherthing.png 2x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/v1/something.com/otherthing.png 3x"
    media="(min-width: 600px)"
  />
  <source
    data-srcset="https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_1.0,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/v1/something.com/otherthing.png 1x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_2.0,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/v1/something.com/otherthing.png 2x, https://res.cloudinary.com/getethos/image/upload/c_fill,dpr_3.0,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/v1/something.com/otherthing.png 3x"
    media="(max-width: 599px)"
  />
  <img
    alt="alt text"
    className="Image blurUp lazyload testImage"
    src="https://res.cloudinary.com/getethos/image/upload/c_fill,f_auto,fl_progressive:semi,q_auto:eco,t_unsupported/v1/something.com/otherthing.png"
    srcSet="https://res.cloudinary.com/getethos/image/upload/c_fill,f_auto,fl_progressive:semi,h_100,q_auto:eco,w_100/t_lqip/v1/something.com/otherthing.png 600w, https://res.cloudinary.com/getethos/image/upload/c_fill,f_auto,fl_progressive:semi,h_200,q_auto:eco,w_200/t_lqip/v1/something.com/otherthing.png 900w, https://res.cloudinary.com/getethos/image/upload/c_fill,f_auto,fl_progressive:semi,h_300,q_auto:eco,w_300/t_lqip/v1/something.com/otherthing.png 1200w, https://res.cloudinary.com/getethos/image/upload/c_fill,f_auto,fl_progressive:semi,h_400,q_auto:eco,w_400/t_lqip/v1/something.com/otherthing.png 1201w"
  />
</picture>
`;
